### DISCUSSION TITLE
AI with Crazyflie

### DISCUSSION THREAD

**Discussion by FallowLand** [bitcraze/discussions#791] [Q&A]
Hello,

I am trying to do some very simple AI with the Crazyflie. My initial thought was to use the flow deck and multi-ranger to detect obstacles and make decisions. But I noticed you have another deck called the AI deck. Because the AI deck is more expensive than the multi-ranger (I already have a flow deck), I wanted to know if the AI deck was overkill. 

Thank you very much for your help!

**Comment by tobbeanton** [comment]
It all depends... Do you want to do the AI part on the PC side? Or perhaps at least start there? Then AI-deck is overkill and just too complex. Later if it runs well and you want to move it on board the Crazyflie you could first try to run it on the STM32F4, but perhaps this will not be powerful enough, then AI-deck might be an option.

**Reply by FallowLand** [reply]
Thank you for the suggestion. 
I have a couple of follow-up questions though:

1. If I don't use the AI deck, can you recommend a camera?
2. Of the 3 options (computer, STM32F34, AI deck), what are the pros and cons (eg. refresh rate, image pixel count, etc.)? 
3. Please refer to this video: https://www.youtube.com/watch?v=CI2Fct8DLbk. Are you aware of this work and do you know who performed it? The title suggested they use an NVIDIA jetson which is too large to carry on the Crazyflie. Can you explain why they didn't use their computer?

Thanks!

**Comment by knmcguire** [comment]
Hi @FallowLand 

I'll provide some help with your questions:

- The princeton robotics course has made their own camera deck which is perhaps a good reference. You can check out [their blogpost here](https://www.bitcraze.io/2022/01/introduction-to-robotics-at-princeton/).  You can also have a search on our old forum for some [build suggestions](https://forum.bitcraze.io/search.php?keywords=camera&sid=805d5c998fe9ee1819ce3f5bf6d28faa)
- The pros and cons is very difficult to determine I'm afraid. Offboard is ofcourse the best but then you are dealing with communication delay, on the stm there is not so much space, but if you keep your obstacle avoidance simple enough it should work. There are more possibilities on the AIdeck, but that is difficult to work with if you are not an embedded system developer. I'll share some papers and some work for you to get perhaps a better idea of what is possible.
- Ah! that's cool! No we are not familiar with this work. Yes a jetson is way too heavy and power hungery, so I think they just used the jetson as an external computer and let it communicate directly with the Crazyflie. I can't see what kind of camera they are using here but most likely their own self-made deck since the AIdeck wasn't released then. 


Here some interesting papers for you to get an idea of what to be able to do with the Crazyflie and navigation:

With the AIdeck:
- ICRA 2023 Work by TU Delft. Estimating optical flow and avoidancing obstacles with the aideck https://www.bitcraze.io/2022/11/nanoflownet-real-time-dense-optical-flow-on-a-nano-quadcopter/
- ICRA 2023 work by University of Bologna. Pose estimation on the aideck: https://arxiv.org/abs/2303.01931

With multiranger onboard the stm:
- There are several blogpost with a collab work of Harvard and Tudelft: [blogpost 1](https://www.bitcraze.io/2019/10/learning-to-seek-autonomous-source-seeking-using-deep-reinforcement-learning-onboard-a-crazyflie/), [ Blogpost 2](https://www.bitcraze.io/2021/07/sniffy-bug-a-fully-autonomous-swarm-of-gas-seeking-nano-quadcopters-in-cluttered-environments/). This work also included an additional light or gas sensor
- Tudelft work for only flowdeck and multiranger and bug algorithms: [blogpost](https://www.bitcraze.io/2019/10/enabling-swarm-exploration/). We have the wallfollowing as a standard [in our examples](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples) for onboard computation.

With multiranger offboard on a computer:
- University of Virginia, doing slam offboard with the multiranger: [IEEE xplore paper page](https://ieeexplore.ieee.org/document/9981200)
- My own ROS project for SLAM offboard with ROS2: [blogpost 1](https://www.bitcraze.io/2022/07/crazyflie-summer-project-with-ros2-and-mapping/) [blogpost 2](https://www.bitcraze.io/2022/08/crazyflie-ros2-summer-update/).

Offboard camera and slam:
- Old paper with FPV camera and offboard slam: http://jsturm.de/publications/data/dunkley14iros.pdf. This was still with Crazyflie 1
- I guess your youtube video is an good example as well! But I haven't seen any other work so far in the last ICRA's and IROS's for offboard visual SLAM (not that I have searched very hard).


Hope that this give you enough info to get started on this!
